### PR TITLE
IaC: Terraform and ACM configurations for Multi-Autopilot GKE environment setup with CI/CD, Anthos & CloudSQL

### PR DESCRIPTION
**Pull request by aablsk**
### Background 
Initially the fork that this PR is based on was created to showcase Google Cloud's CICD story encompassing Cloud Build / Cloud Deploy / kustomize / skaffold. In order to maximize value from the work done for the showcase we decided to automate the setup with terraform utilising Cloud Foundation Toolkit.

### Change Summary
Added configuration that includes:
- Terraform for all infrastructure (with exception of Github repo connection which is a one-time manual step)
- CI per team
  - CloudBuild /w skaffold (includes caching)
- CD (staging, production) per team
  - CloudDeploy /w skaffold profiles per environment
- 3 Environments:
  - Development (skaffold dev only)
    - GKE 
    - Autopilot
    - ASM
    - ACM for service accounts, namespaces, rolebindings…
    - Workload identity
    - Namespace: bank-of-anthos-development
  - Staging (CD deploy)
    - GKE 
    - Autopilot
    - ASM
    - ACM for service accounts, namespaces, rolebindings…
    - Workload identity
    - Namespace: bank-of-anthos-staging
    - CloudSQL
  - Production (CD deploy)
    - GKE 
    - Autopilot
    - ASM
    - ACM for service accounts, namespaces, rolebindings…
    - Workload Identity
    - Namespace: bank-of-anthos-production
    - CloudSQL
- Heavy use of kustomize components to stay DRY
- Minimized service account permissions
- Uses CloudFoundation Toolkit for Kubernetes (which restricts modularization of environments due to kubernetes provider circular dependency)

### Additional Notes
This PR does NOT include the changes to the src/ folder, which include kustomizations for microservices, skaffold profiles and cloudbuild.yaml definitions for individual teams. 
The terraform code included in this PR does however reference these files, which leads to this IaC not being immediately compatible with the current application source.

### Testing Procedure
Follow steps [here](https://github.com/aablsk/bank-of-anthos/blob/main/iac/tf-multienv-cicd-anthos-autopilot/README.md#how-to-set-it-up) using the prepared shell scripts for easiest setup.

Please note that script 4 and 5 will log errors as they require the second PR (which includes changes to src/ folder)

### Related PRs or Issues 
I stole shamelessly from these PRs:
https://github.com/GoogleCloudPlatform/bank-of-anthos/pull/881
https://github.com/GoogleCloudPlatform/bank-of-anthos/pull/792

Changes to src/ can be seen in this [fork](https://github.com/aablsk/bank-of-anthos)

### CODE REVIEW

**Comment by aablsk** [comment]
@bourgeoisor What's our requirement on empty lines at end of file? (need to fix this still)

File: iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/03-wait-for-asm.sh
Comment (bourgeoisor) [inline iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/03-wait-for-asm.sh:18]: TF doesn't do that waiting? :(

File: iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/04-init-cloud-sql.sh
Comment (bourgeoisor) [inline iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/04-init-cloud-sql.sh:1]: This script ends up replacing: https://github.com/GoogleCloudPlatform/bank-of-anthos/blob/main/extras/cloudsql/create_cloudsql_instance.sh correct?

Comment (bourgeoisor) [reply iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/04-init-cloud-sql.sh:1]: Will there still be a non-TF-friendly way of deploying with Cloud SQL through kustomize / skaffold alone?

File: iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/00-setup-all-the-things.sh
Comment (bourgeoisor) [inline iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/00-setup-all-the-things.sh:1]: I wonder if another name / something less ambiguous would be more suitable, like `00-run-scripts-01-through-05` or ``00-all-scripts` or something...

File: iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/01-patch-acm-config.sh
Comment (bourgeoisor) [inline iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/01-patch-acm-config.sh:1]: I'm not super sure about this script -- I think I'd prefer if we'd hardcode (or default a variable) to this current GCP project / region and not have this script at all. I'd rather `git` commands (besides `git clone`) not be involved.

Comment (bourgeoisor) [reply iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/01-patch-acm-config.sh:1]: Perhaps instead of a script, it could be part of the README, where in the prereq you state that if you're deploying this TF outside the context of BoA maintainers, you should:
- fork the repo
- substitute some stuff like the GCP project name and commit those in your now forked repo (perhaps that bit can be a quick script, but it should not be part of the TF family of script) -- though I'd personally lean towards simplifying and not scripting it at all... Just stating it in the README
- THEN you can run the TF scripts

File: iac/tf-multienv-cicd-anthos-autopilot/README.md
Comment (bourgeoisor) [inline iac/tf-multienv-cicd-anthos-autopilot/README.md:1]: Even before this, I'd give a disclaimer saying that this entire TF directory is not meant for everyday usage and is _not_ required to deploy Bank of Anthos or any of its extensions / flavours. It's meant to showcase how you could do a full CI/CD environment, etc etc...

File: iac/tf-multienv-cicd-anthos-autopilot/main.tf
Comment (bourgeoisor) [inline iac/tf-multienv-cicd-anthos-autopilot/main.tf:19]: Commented out?

File: iac/tf-multienv-cicd-anthos-autopilot/env-staging.tf
Comment (bourgeoisor) [inline iac/tf-multienv-cicd-anthos-autopilot/env-staging.tf:35]: Commented out?

File: iac/tf-multienv-cicd-anthos-autopilot/README.md
Comment (bourgeoisor) [inline iac/tf-multienv-cicd-anthos-autopilot/README.md:44]: This means BoA gets fully deployed as well, correct? Does it get these things? https://github.com/GoogleCloudPlatform/bank-of-anthos/tree/main/.github/prod-cluster

Ideally we'd dogfood this entire TF flow for real when we develop on BoA and the prod cluster would be the https://bank-of-anthos.xyz/ cluster :)

**Review by bourgeoisor — CHANGES_REQUESTED** [review]
Thank you Alex for the PR, added a handful of comments and questions :)

File: iac/acm-multienv-cicd-anthos-autopilot/base/kustomization.yaml
Comment (aablsk) [reply iac/acm-multienv-cicd-anthos-autopilot/base/kustomization.yaml:1]: [Yes it does!](https://cloud.google.com/anthos-config-management/docs/tutorials/config-sync-helm) ✨

File: iac/acm-multienv-cicd-anthos-autopilot/base/config.yaml
Comment (aablsk) [reply iac/acm-multienv-cicd-anthos-autopilot/base/config.yaml:1]: I propose to 'move' these files instead of 'copying' them. Since this base configuration is a kustomize component, it can be easily added to a skaffold configuration. As discussed I have prepared a skaffold profile ('development') which is compatible with a vanilla k8s cluster for the next PR that will include this. 

Source to include these manifests specifically can be found [here](https://github.com/aablsk/bank-of-anthos/blob/94777354f75cc90c7e19da45f0e640d1b40b3bac/skaffold.yaml#L14-L24).

File: iac/tf-multienv-cicd-anthos-autopilot/variables.tf
Comment (aablsk) [reply iac/tf-multienv-cicd-anthos-autopilot/variables.tf:17]: We could do this, but I'm more leaning towards not doing this because:
- this is not a multi-cloud repo, which means there is little value in adding this prefix
- CFTK terraform modules always use `project_id`, terraform modules use `project` as the key for this value e.g.

```
module "enabled_google_apis" {
  ...
  project_id                  = var.project_id
  ...
}
resource "google_artifact_registry_repository" "container_registry" {
  ...
  project       = var.project_id
  ...
}

```

If you still prefer `google_cloud_project_id`, I'm happy to change it.

File: iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/03-wait-for-asm.sh
Comment (aablsk) [reply iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/03-wait-for-asm.sh:18]: TF is completely done after 02-terraform.sh

As these scripts' purpose is mainly to make it easier for me to test reproducibility and it seems they are source for quite a bit of confusion, I'm removing the entire script folder. The necessary one-time setup steps will instead be moved into README.md.

File: iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/04-init-cloud-sql.sh
Comment (aablsk) [reply iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/04-init-cloud-sql.sh:1]: > This script ends up replacing: https://github.com/GoogleCloudPlatform/bank-of-anthos/blob/main/extras/cloudsql/create_cloudsql_instance.sh correct?
Not as far as I can tell. The linked script is setting up the infrastructure (CloudSQL instances). This is done entirely in terraform in this PR ([staging](https://github.com/aablsk/bank-of-anthos/blob/2b6f6f77b6e2d116bc705f696efc274515035df9/iac/tf-multienv-cicd-anthos-autopilot/env-staging.tf#L70-L99), [production](https://github.com/aablsk/bank-of-anthos/blob/2b6f6f77b6e2d116bc705f696efc274515035df9/iac/tf-multienv-cicd-anthos-autopilot/env-production.tf#L70-L98])). The purpose of `04-init-cloud-sql.sh` is to [deploy the kubernetes jobs](https://github.com/GoogleCloudPlatform/bank-of-anthos/tree/main/extras/cloudsql/populate-jobs) that will initialise the CloudSQL instances with data.

> Will there still be a non-TF-friendly way of deploying with Cloud SQL through kustomize / skaffold alone?
There is currently no way to deploy CloudSQL through kustomize/skaffold alone. Instead what we rely on to provision infrastructure is the shell script you linked which heavily uses gcloud cli. This is still possible with the proposed solution. It could be the possibility to provision with kustomize + skaffold, but this would imply usage of [config connector](https://cloud.google.com/config-connector/docs/overview) and is out of scope for this PR.

File: iac/tf-multienv-cicd-anthos-autopilot/env-staging.tf
Comment (aablsk) [reply iac/tf-multienv-cicd-anthos-autopilot/env-staging.tf]: Removed commented line.

File: iac/tf-multienv-cicd-anthos-autopilot/variables.tf
Comment (aablsk) [reply iac/tf-multienv-cicd-anthos-autopilot/variables.tf:53]: > What is the purpose of this variable? Is it for dev purposes? Or if someone forks the repo indefinitely? If the latter, could this be a hardcoded value instead?

This is for forked repositories. Otherwise we could make this a terraform `local` variable and hardcode `GoogleCloudPlatform` as the owner.

> Actually, could we not merge "sync_repo" and "repo_owner" together such that the variable value would look like "aablsk/bank-of-anthos" or even better, the full URL: "https://www.github.com/aablsk/bank-of-anthos"

Unfortunately that would complicate things as the terraform `google_cloudbuild_trigger` resource expects the `owner` value ([example](https://github.com/aablsk/bank-of-anthos/blob/fb155665f3941afa22e495ee7afb21bdeb9fb31e/iac/tf-multienv-cicd-anthos-autopilot/modules/ci-cd-pipeline/cloud-build-trigger.tf#L19-L26)). I could solve this using the [terraform string split manipulation](https://developer.hashicorp.com/terraform/language/functions/split), if you prefer using the repo url.

File: iac/tf-multienv-cicd-anthos-autopilot/README.md
Comment (aablsk) [reply iac/tf-multienv-cicd-anthos-autopilot/README.md:4]: Done.

File: iac/acm-multienv-cicd-anthos-autopilot/overlays/development/kustomization.yaml
Comment (aablsk) [reply iac/acm-multienv-cicd-anthos-autopilot/overlays/development/kustomization.yaml]: This was previously replaced by the [01-patch-acm-config.sh](https://github.com/aablsk/bank-of-anthos/blob/private-clusters/iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/01-patch-acm-config.sh) script. As a proposal I have removed all bash scripts and instead added step-by-step instructions for the initial (one-time) setup to the README.md

File: iac/acm-multienv-cicd-anthos-autopilot/overlays/production/configmap.yaml
Comment (aablsk) [reply iac/acm-multienv-cicd-anthos-autopilot/overlays/production/configmap.yaml]: This was previously replaced by the [01-patch-acm-config.sh](https://github.com/aablsk/bank-of-anthos/blob/private-clusters/iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/01-patch-acm-config.sh) script. As a proposal I have removed all bash scripts and instead added step-by-step instructions for the initial (one-time) setup to the README.md

File: iac/acm-multienv-cicd-anthos-autopilot/overlays/production/kustomization.yaml
Comment (aablsk) [reply iac/acm-multienv-cicd-anthos-autopilot/overlays/production/kustomization.yaml]: This was previously replaced by the [01-patch-acm-config.sh](https://github.com/aablsk/bank-of-anthos/blob/private-clusters/iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/01-patch-acm-config.sh) script. As a proposal I have removed all bash scripts and instead added step-by-step instructions for the initial (one-time) setup to the README.md

File: iac/acm-multienv-cicd-anthos-autopilot/overlays/staging/configmap.yaml
Comment (aablsk) [reply iac/acm-multienv-cicd-anthos-autopilot/overlays/staging/configmap.yaml]: This was previously replaced by the [01-patch-acm-config.sh](https://github.com/aablsk/bank-of-anthos/blob/private-clusters/iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/01-patch-acm-config.sh) script. As a proposal I have removed all bash scripts and instead added step-by-step instructions for the initial (one-time) setup to the README.md

File: iac/acm-multienv-cicd-anthos-autopilot/overlays/staging/kustomization.yaml
Comment (aablsk) [reply iac/acm-multienv-cicd-anthos-autopilot/overlays/staging/kustomization.yaml]: This was previously replaced by the [01-patch-acm-config.sh](https://github.com/aablsk/bank-of-anthos/blob/private-clusters/iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/01-patch-acm-config.sh) script. As a proposal I have removed all bash scripts and instead added step-by-step instructions for the initial (one-time) setup to the README.md

File: iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/env.sh
Comment (aablsk) [reply iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/env.sh]: As a proposal I have removed all bash scripts and instead added step-by-step instructions for the initial (one-time) setup to the README.md.

File: iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/00-setup-all-the-things.sh
Comment (aablsk) [reply iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/00-setup-all-the-things.sh]: As a proposal I have removed all bash scripts and instead added step-by-step instructions for the initial (one-time) setup to the README.md

File: iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/01-patch-acm-config.sh
Comment (aablsk) [reply iac/tf-multienv-cicd-anthos-autopilot/shell-scripts/01-patch-acm-config.sh]: I agree. These scripts were mostly created to support me in testing reproducibility of the setup with little effort. For this PR I have therefore removed all bash scripts and instead added step-by-step instructions for the initial (one-time) setup to the README.md.

File: iac/tf-multienv-cicd-anthos-autopilot/main.tf
Comment (aablsk) [reply iac/tf-multienv-cicd-anthos-autopilot/main.tf:19]: I have removed the comment and added instructions for setting this up to the README.md

File: iac/tf-multienv-cicd-anthos-autopilot/README.md
Comment (aablsk) [reply iac/tf-multienv-cicd-anthos-autopilot/README.md]: Agreed. I was not aware of this configuration, but happy to add it with the second PR. (will include terraform for CloudArmor, IP... and adapted production skaffold profile with a fitting kustomization)

**Comment by aablsk** [comment]
@bourgeoisor Could you please provide me with the following values that you'd like to use for your setup, so I can include them in the configuration:
1. Google Cloud ProjectId in which you are planning to deploy this
2. Region in which the resources should be deployed (if applicable)
3. Name of the GCS bucket in which you'd like to store the shared terraform state

Thanks in advance!

File: iac/tf-multienv-cicd-anthos-autopilot/apis.tf
Comment (NimJay) [inline iac/tf-multienv-cicd-anthos-autopilot/apis.tf:39]: **Praise:** I like that we're dedicating a separate `apis.tf` file for GCP API enablement. This is clean!

File: iac/tf-multienv-cicd-anthos-autopilot/artifact-registry.tf
Comment (NimJay) [inline iac/tf-multienv-cicd-anthos-autopilot/artifact-registry.tf:30]: **Thought:** After merging this pull-request, we should consider tracking our desire to use a custom service account here in a GitHub issue (i.e., create a new GitHub Issue to add visibility).

File: iac/tf-multienv-cicd-anthos-autopilot/apis.tf
Comment (aablsk) [reply iac/tf-multienv-cicd-anthos-autopilot/apis.tf:39]: 🙈 Thanks for the kind words, Nim!

File: iac/tf-multienv-cicd-anthos-autopilot/artifact-registry.tf
Comment (aablsk) [reply iac/tf-multienv-cicd-anthos-autopilot/artifact-registry.tf:30]: Agreed!

**Review by bourgeoisor — APPROVED** [review]
LGTM

**Pull request merged**
